### ISSUE TITLE
Nullable reference types are unsupported

### ISSUE THREAD

**Issue by bddckr** [dotnet/orleans#5834]
C# 8.0 is still in preview, I know... We're getting closer to the release of .NET Core 3.0 and C# 8.0, though!

I'm using C# 8 and enabled `Nullable`. Now hitting a small roadblock while designing some Grain API: I use nullable reference types as part of the Grain interfaces to ensure me and my colleagues have a codified contract with regard to the nullability of that API's parameters and its return value.

The MSBuild codegen of Orleans results in lines like this:
```c#
context.SerializeInner(input.Arguments, typeof(string? ));
```
(Unrelated: The space after the question mark - strange quirk all nullable reference types have in the generated code.)

And turns out while that is still marked as an open question https://github.com/dotnet/csharplang/issues/2201 for the lang team, the compiler team went ahead and resolved it, by making `typeof` of a nullable reference type a compiler error https://github.com/dotnet/roslyn/pull/35001.

Can Orleans instead just work with `typeof(T)` instead of `typeof(T?)` in this auto-generated code? As noted in the compiler PR above this basically is the same type in the type system anyway:
> The reason is for this apparent inconsistency is that it's actually in line with previous decisions. The root problem we're trying to address here is that parts of the C# type system aren't reflected in the runtime type system. Hence `typeof(List<string?>)` and `typeof(List<string>)` will have the same value.
> https://github.com/dotnet/roslyn/issues/29894#issuecomment-482274990

@ReubenBond was so kind and already chimed in on gitter:
> @bddckr I was asking myself yesterday "how long until someone opens an issue on nullable reference type support?" 😜
> 
> Hopefully it's simple enough to fix (would you like to take a stab at it?), but I anticipate a few issues. The notnull type constraint will be another
> https://gitter.im/dotnet/orleans?at=5d4b1564d03a7c63e6e81cc8

> @bddckr if for example the user creates some container type, `MyContainer<T> where T : notnull` there may be some requirement for generated code to also include nullability assertions. It may be easier to just disable nullable reference types for the generated files, but it's possible that the user would have generic grain methods or grain interfaces which have those `notnull` constraints, so in that case we likely would need to copy the constraints onto the generated classes (GrainReference impl, for ex). We already do that for other constraints. I think there are other new constraints which we need to handle (`enum`?)
> https://gitter.im/dotnet/orleans?at=5d4b18fc2612bb718c2862de

I don't think the _attributes_ would change much really - Orleans is only generating this for serialization AFAICS. I don't believe any serialization in C# was updated yet for nullable ref types, so we might just get away with it as they'll all do...

`notnull` as mentioned by Reuben is actually special - it's a language keyword, allowing a type parameter to be constrained.

---

I'm actively trying to work on this and will be reporting back in here. Hopefully it won't be too much work 🤞

**Comment by bddckr** [comment]
The issue of generating invalid `typeof` instructions is fixed by upgrading the `Microsoft.CodeAnalysis.CSharp` package.

I actually ended up using `3.3.0-beta3-19407-05` (from [Roslyn's MyGet feed](https://dotnet.myget.org/feed/roslyn/package/nuget/Microsoft.CodeAnalysis.CSharp)) in my attempt to get this handled. https://github.com/dotnet/roslyn/pull/37301 is needed to ensure the existing usage of `SwitchStatement` in Orleans' code generator does actually include the needed parenthesis around the expression (see the fixed issue https://github.com/dotnet/roslyn/issues/33685 for more details). Other than that everything just works as the `typeof` operator now is generated with the reference type that is used as a nullable one as the argument - it's one and the same type to the type system after all (see my initial comment).

I'm not sure how to move forward with this as I think we'd all prefer close to no changes to Orleans' code generator source if we can help it. On the other hand this currently needs a preview package version to be referenced... From a quick check nothing else seems to break at least.

---

I didn't tackle handling the nullability annotation attributes or the `notnull` constraint yet. These are definitely more in the category of "nice to have" rather than being necessary to unblock users of nullable reference types and Orleans.

My knowledge of the Roslyn APIs is very limited so perhaps we reduce the scope of this issue down to be about unblocking users of nullable reference types. Copying the attributes and constraints to the generated types could be one or two additional issues to tackle later.

**Comment by ReubenBond** [comment]
Thank you, @bddckr. Let's schedule this for when those packages have a stable release. Hopefully that coincides with .NET Core 3 (Sept 23) or sooner

**Comment by erikljung** [comment]
Found another work around for this issue, it's hacky but is doable without forking Orleans.
**This code is not tested or verified in production**

_SomeLib.csproj (Lib which generates code -- holds Grain interfaces)_
```
<Project Sdk="Microsoft.NET.Sdk">
   <Import Project="FixGeneratedCode.targets" />
   <Target Name="FixGeneratedCode" AfterTargets="OrleansGenerateCode">
      <FixGeneratedCode Path="$(IntermediateOutputPath)$(AssemblyName).orleans.g.cs"/>
   </Target>
```

_FixGeneratedCode.targets_
```
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <UsingTask
    TaskName="FixGeneratedCode"
    TaskFactory="RoslynCodeTaskFactory"
    AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
    <ParameterGroup>
      <Path ParameterType="System.String" Required="true"/>
    </ParameterGroup>
    <Task>
      <Using Namespace="System"/>
      <Using Namespace="System.IO"/>
      <Code Type="Fragment" Language="cs">
<![CDATA[
var content = File.ReadAllText(Path);
content = content.Replace("? )", ")");
File.WriteAllText(Path, content);
]]>
      </Code>
    </Task>
  </UsingTask>
</Project>
```

**Comment by bddckr** [comment]
Ha I love that! Thanks a ton for sharing - I'm using it from here on instead of doing my own forked bugfix.

**Comment by erikljung** [comment]
Glad you liked it! For the record this wizardry was done by my colleague @palpha

**Comment by palpha** [comment]
I strongly suggest tweaking this approach to be more precise in what it changes. This is just a proof of concept, even though it seems to work well for now :).

**Comment by bddckr** [comment]
True! This works fine for me:
```c#
var content = File.ReadAllText(Path);
content = "#nullable enable\n"
    + "#pragma warning disable nullable\n"
    + "#pragma warning disable 8603\n"
    + content
    + "#pragma warning restore nullable\n"
    + "#nullable restore\n";
File.WriteAllText(Path, content);
```
